orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -300,6 +522,130 @@ Status TransformGraphForPipeline(
     backward_waited_event_name,
     forward_recorded_output_name,
     backward_waited_output_name));
+  
+  // Different stages have different patterns of Send & Recv.
+  // For different patterns, we add different WaitEvent and Record.
+  //
+  // After applying all transformations above, the current patterns
+  // are listed below.
+  //
+  // 1. First stage:
+  //  Wait -----------------> FW -> --------> Send -> Record ->
+  //  Wait -> Recv ---------> BW -> ----------------> Record
+  // 2. Middle stage:
+  //  Wait -> Recv -> Wait -> FW -> Record -> Send -> Record ->
+  //  Wait -> Recv -> Wait -> BW -> Record -> Send -> Record
```

Comment:
the wait op after recv and record op before send is not inserted yet at this point?

Reply:
They are not. One line in this comment says:
```
  // After applying all transformations below, we will have
  // the following patterns.
```
This big chunk of comments aims at explaining status before and after the following transforms.

Reply:
right, agree the patterns describing the 'after' transform is correct. I was wondering the comment on line 550-551, which describes the 'before' pattern. In 'before' pattern, no wait op after recv nor record op before send is inserted. Maybe we should correct this in the comment?

---
In reply to: [425964057](https://github.com/microsoft/onnxruntime/pull/3823#discussion_r425964057) [](ancestors = 425964057)

Reply:
The first 3 graphs is for `status after applying previous transformations`. The later 3 graphs is for `status after applying all transformation`.

---
In reply to: [426081946](https://github.com/microsoft/onnxruntime/pull/3823#discussion_r426081946) [](ancestors = 426081946,425964057)